lib/bamboo/view.ex

```diff
@@ -0,0 +1,201 @@
+      |> Enum.map(&template_and_name/1)
+      |> Enum.map(fn {template, name} ->
+        {name, compile(template, name)}
+      end)
+
+    template_names = Enum.map(results, &elem(&1, 0))
+    render_functions = Enum.map(results, &elem(&1, 1))
+
+    quote do
+      unquote(render_functions)
+
+      def render_template(template, _assigns) do
+        raise UndefinedTemplateError,
+          module: __MODULE__,
+          template: template,
+          available: unquote(template_names)
+      end
+    end
+  end
+
+  @doc false
+  def render_email(view, email, template, assigns) do
+    email
+    |> put_defaults(view)
+    |> merge_assigns(assigns)
+    |> render(template)
+  end
+
+  defp put_defaults(%{private: private} = email, view_module) do
+    private =
+      private
+      |> Map.put_new(:html_layout, false)
+      |> Map.put_new(:text_layout, false)
+      |> Map.put_new(:view_module, view_module)
+
+    %{email | private: private}
+  end
+
+  defp merge_assigns(%{assigns: email_assigns} = email, assigns) do
+    assigns = email_assigns |> Map.merge(Enum.into(assigns, %{}))
+    email |> Map.put(:assigns, assigns)
+  end
+
+  defp render(email, template) when is_atom(template) do
+    render_html_and_text_emails(email, template)
+  end
+
+  defp render(email, template) do
+    render_text_or_html_email(email, template)
+  end
+
+  defp render_html_and_text_emails(email, template) do
+    view_template = Atom.to_string(template)
+
+    email
+    |> Map.put(:html_body, render_html(email, view_template <> ".html"))
+    |> Map.put(:text_body, render_text(email, view_template <> ".text"))
+  end
+
+  defp render_text_or_html_email(email, template) do
+    cond do
+      String.ends_with?(template, ".html") ->
+        email |> Map.put(:html_body, render_html(email, template))
+
+      String.ends_with?(template, ".text") ->
+        email |> Map.put(:text_body, render_text(email, template))
+
+      true ->
+        raise ArgumentError, """
+        Template name must end in either ".html" or ".text". Template name was #{
+          inspect(template)
+        }
+
+        If you would like to render both and html and text template,
+        use an atom without an extension instead.
+        """
+    end
+  end
+
+  defp render_html(email, template) do
+    layout = email.private.html_layout
+    render_within_layout(layout, email, template)
+  end
+
+  defp render_text(email, template) do
+    layout = email.private.text_layout
+    render_within_layout(layout, email, template)
+  end
+
+  defp render_within_layout(_layout = false, email, template) do
+    module = email.private.view_module
+    module.render_template(template, email.assigns)
+  end
+
+  defp render_within_layout({layout_view, layout_template}, email, template) do
+    module = email.private.view_module
+    contents = module.render_template(template, email.assigns)
+
+    assigns =
+      email.assigns
+      |> Map.put(:inner_content, contents)
```

Comment:
```suggestion
      Map.put(email.assigns, :inner_content, contents)
```

Gotta keep up the fight now that I started it! :)